src/main/scala/io/getclump/ClumpContext.scala

```diff
@@ -7,31 +7,47 @@ private[getclump] final class ClumpContext {
   private[this] val fetchers =
     new HashMap[ClumpSource[_, _], ClumpFetcher[_, _]]()
 
-  def flush(clumps: List[Clump[_]]): Future[Unit] =
+  def flush(clumps: List[Clump[_]]): Future[Unit] = {
+    // 1. Get a list of all visible clumps grouped by level of composition
+    val clumpsByLevel = getClumpsByLevel(clumps)
+
+    // 2. Flush the fetches from all the visible clumps
+    flushFetches(clumpsByLevel.flatten).flatMap { _ =>
+      // 3. Walk through the downstream clumps as well, starting at the deepest level
+      flushDownstreamByLevel(clumpsByLevel.reverse)
+    }
+  }
+
+  // Unfold all visible (ie. upstream) clumps
+  private[this] def getClumpsByLevel(clumps: List[Clump[_]]): List[List[Clump[_]]] = {
     clumps match {
+      case Nil => Nil
+      case _ => clumps :: getClumpsByLevel(clumps.flatMap(_.upstream))
+    }
+  }
+
+  private[this] def flushDownstreamByLevel(levels: List[List[Clump[_]]]): Future[Unit] = {
+    levels match {
       case Nil => Future.successful(())
-      case _ =>
-        flushUpstream(clumps).flatMap { _ =>
-          flushFetches(clumps).flatMap { _ =>
-            flushDownstream(clumps)
+      case head :: tail =>
+        // 1. Resolve the downstream clumps (will succeed because deeper downstream clumps have already been resolved)
+        Future.sequence(head.map(_.downstream)).flatMap { res =>
+          // 2. Flush the resulting clumps, thus completely resolving this subtree of the composition
+          flush(res.flatten).flatMap { _ =>
+            // 3. Move up one level of composition and repeat this process
+            flushDownstreamByLevel(tail)
           }
         }
     }
+  }
 
-  private[this] def flushUpstream(clumps: List[Clump[_]]) =
-    flush(clumps.map(_.upstream).flatten)
-
-  private[this] def flushDownstream(clumps: List[Clump[_]]) =
-    Future.sequence(clumps.map(_.downstream)).flatMap { down =>
-      flush(down.flatten.toList)
-    }
-
+  // Flush all the ClumpFetch instances in a list of clumps, calling their associated fetch functions in parallel if possible
   private[this] def flushFetches(clumps: List[Clump[_]]) = {
```

Comment:
somehow it wasn't obvious to me earlier that this indeed does stuff in parallel, should we rename to `flushFetchesInParallel`? I know you didn't add this message but since we're here :)